Add _repr_mimebundle_ for IPython.display

boeddeker opened the pull request:
Address #138

@xflr6 Thank you for the modifications of the graphviz code base.
I followed your suggestion and implemented the `_repr_mimebundle_ `.
For SVG, PNG and JPEG it works, but PDF and JSON didn't work.
In the IPython documentation is not much info, so I was not able to figure out, why they don't work.
(For PDF, they may expect a file)

I didn't change the name for `JupyterSvgIntegration`, because I wanted to wait for feedback.

xflr6 reviewed (commented):
Thanks for the PR.

 Implementation looks good in general. :)

- I think `_repr_mimebundle_()` does the right thing. Proposed some simplifications
- We will need to move the code in place a bit and make it more consistent with other parts, see my suggestions. Hope this helps.
- Can you add type annotations?

Then there is the whole side of testing and  documentation: It would be great if you can try to take a look at the tests and the documentation (see also https://graphviz.readthedocs.io/en/latest/development.html) and at least make sure no tests are broken and there is at least some code coverage of the added (non-module-level) code.But if you prefer me to take over, that is fine as well. Feel free to propose documentation and changelog entry. Maybe it would make sense to add a docstring and include the method in the API docs (treat it as pseudo-private), WDYT?

Please check flake8 in tests result (when they pass again).

xflr6 commented on graphviz/jupyter_integration.py:
```diff
@@ -1,12 +1,70 @@
+    # def _repr_svg_(self):
+    #     return self.pipe(format='svg', encoding=self._encoding)
+
+    def _repr_mimebundle_(self, include=None, exclude=None, **kwargs):
```
nit: Let's `del` kwargs in the first line explicitly to make explicit we intentionally ignore it (or use `**_):`).

boeddeker commented:
> Then there is the whole side of testing and documentation: It would be great if you can try to take a look at the tests and the documentation (see also https://graphviz.readthedocs.io/en/latest/development.html) and at least make sure no tests are broken and there is at least some code coverage of the added (non-module-level) code.But if you prefer me to take over, that is fine as well. Feel free to propose documentation and changelog entry. Maybe it would make sense to add a docstring and include the method in the API docs (treat it as pseudo-private), WDYT?

I will check the `development` side and add some tests. I skipped them, because I was not sure, if my suggestion is ok.

I have a question regarding the parameters folder. Should I create a new file in `parameters` and write code similar to the other files there? Do you have a suggestion for the name? `jupyter_representation.py` and `class JupyterRepresentation: ...` is lengthy and `juypter` alone not informative.

Btw. when you think it takes less time for you, when you take over instead of suggesting changes, feel free to do so.
(I enabled "Allow edits by maintainers", so you can apply changes directly to this PR.)

> Can you add type annotations?

Sure. 

> Please check flake8 in tests result (when they pass again).

Sure, I try to get all test working, including the style checks.

boeddeker commented on graphviz/jupyter_integration.py:
```diff
@@ -1,12 +1,70 @@
 
+_default_jupyter_representation = 'image/svg+xml'
+
+_jupyter_repr_functions = {
+    'image/svg+xml': lambda graph: graph.pipe(format='svg',
+                                              encoding=graph._encoding),
+    'image/png': lambda graph: graph.pipe(format='png'),
+    'image/jpeg': lambda graph: graph.pipe(format='jpeg'),
+
+    # The following does not work yet. Not clear why, because the documentation
+    # has only few information about them.
+    # 'application/json': lambda graph: json.loads(
+    #     graph.pipe(format='json', encoding=graph._encoding)),
+    # 'application/pdf': lambda graph: graph.pipe(format='pdf'),
+
+    # No reasonable representation
+    # 'text/plain': None,
+    # 'text/latex': None,
+    # 'text/html': None,
+    # 'application/javascript': None,
+    # 'text/markdown': None,
+}
+
+JUPYTER_REPRESENTATIONS = set(_jupyter_repr_functions.keys())
```
If I get it right, the methods will later be on the graph object, so they should be verbose, i.e. `_repr_image_svg_xml`.
When you see that, you will see the similarity with the suggestion from IPython `_repr_svg_` and it would then be more natural to use `_repr_svg_`. Sadly, the existence of `_repr_svg_` will disable `_repr_mimebundle_`.

So the private methods will differ from the suggested ones, and that feels slightly strange.

I just wanted to clarify this. With a "global" dict, it would also be slightly easier to add custom reprs (I could add the PDF to svg example to the docs, which I don't want to add to the source code, because it probably only work on linux).

xflr6 replied:
As these methods are properly private/internal (as oposed to `__repr_svg_` and `_repr_mimebundle_` which are pseudo-private), I think the naming is really only relevant for developers of this library (not for end users).

`_repr_image_svg_xml` SG.

>  With a "global" dict

Not sure if I understand what exactly you mean with this, can you expand a bit?

In case your goal is to allow end users to implement formats for which we don't provide a default implementation I guess we would need to change the logic for putting together `mimebundle` above to ` if format in reprs and hasattr(self, method_name)`. However, I am not convinced yet that we should support that.

boeddeker replied:
> Not sure if I understand what exactly you mean with this, can you expand a bit?

My motivation to write the issue was, that I had a broken graphviz installation and wanted to change, how jupyter displays the graphs. Since it was for teaching, I searched for a solution inside a Jupyter Notebook. e.g. put
```python
import graphviz
def display_fix(...): ...
graphviz.jupyter_integration.JUPYTER_REPRESENTATIONS['graphviz_vX.Y.Z_fix'] = display_fix()
graphviz.jupyter_integration.set_default_jupyter_representation('graphviz_vX.Y.Z_fix')
```
in the first cell. Now where you mention it
```python
import graphviz
def display_fix(...): ...
graphviz.jupyter_integration.JupyterSvgIntegration._repr_mimebundle_ = display_fix
```
will also do the job. More hacky, but does job.

I will follow you suggestion.

xflr6 replied:
I see, thanks.

If one would need to monkeypatch, I think one could also do it like this:

```python
class JupyterIntegrationPatch:
    def _repr_mimebundle_(sef, ...):
          ...

 class Graph(JupyterIntegrationPatch, graphviz.Graph):
    ...

 class Digraph(JupyterIntegrationPatch, graphviz.Digraph):
    ...

 class Source(JupyterIntegrationPatch, graphviz.Source):
    ...
```

boeddeker replied:
That depends, how much control you have. When you use code, that uses `graphviz`, you cannot replace the classes.
In my case, I just want to have a notebook cell, where I modify, what `graphviz` uses for display in a single cell and don't want to touch the external code.

But with the current implementation, I am able to use a workaround for the bug. So its fine for me.

xflr6 replied:
Agreed.

Just to make sure we are on the same page here: when you do something like below, you are monkey-patching private internals of `graphviz`, so if this breaks (e.g. because we rearrange the code or rename internal classes), it's completely your own responsibility (i.e. we won't consider this a bug in `graphviz`).

`graphviz.jupyter_integration.JupyterSvgIntegration._repr_mimebundle_ = display_fix`

boeddeker replied:
Yes, such a code is hacking the internals and may break with a new release. But it is a fix for graphviz (i.e. not Python graphviz) installed on the Linux machine, that I cannot update. So such a hack is in my mind ok and does not need public an API.

xflr6 replied:
Kewl, thanks.

boeddeker commented on graphviz/jupyter_integration.py:
```diff
@@ -1,12 +1,70 @@
+    global _default_jupyter_representation
+
+    if jupyter_representation not in JUPYTER_REPRESENTATIONS:
+        similar = {
+            r
+            for r in JUPYTER_REPRESENTATIONS
+            if jupyter_representation.lower() in r}
+        if similar:
+            raise ValueError(
+                f'unknown jupyter representation: {jupyter_representation!r}\n'
+                f'Suggestions: {similar}')
+        else:
```
Sure, I can make it case-insensitive, also I don't see here a motivation, because the `mime` names aren't obvious.
My idea here was, that the user might request `png` and then gets the suggestion, that the correct values would be `image/png` or in case of svg you need the xml postfix (`image/svg+xml`).

(Normally I use the difflib as in [Dispatcher](https://github.com/fgnt/paderbox/blob/master/paderbox/utils/mapping.py) to get informative exceptions. It saved me much time for debugging.)

xflr6 replied:
On a second thought, maybe case-insensitive would already be overkill.

Supporting `image/svg` as an alias for `image/svg+xml' might help if something in Jupyter changes.

> My idea here was, that the user might request `png`

Just in case: In my understanding this method is not to be called by users.

xflr6 replied:
Sorry for my confusion: the function`set_default_format` is indeed for users as you say. I was thinking about `_repr_mimebundle_`: providing hints SG for the former (if you think its worth implementing) but not for the latter is what I meant to say (e.g. we could call the `verify` function in the former and then provide hints in that function on ValueError). HTH

boeddeker replied:
Ah, ok. Nevertheless, your suggestion is more common in python.

xflr6 commented:
> I skipped them, because I was not sure, if my suggestion is ok.
> when you think it takes less time for you, when you take over instead of suggesting changes, feel free to do so.

All good. I am fine with making suggestions for now (can take over when one/both of us think it's time).

> Should I create a new file in `parameters` and write code similar to the other files there?

I would say no: this submodule is stricly for parameters for rendering:

https://github.com/xflr6/graphviz/blob/78205195b79904ff82fa29a2df541aa7cadbfc69/graphviz/parameters/__init__.py#L1

boeddeker commented:
I tried now to be as close as possible to the code you mentioned.

In `tests/test_parameters.py` I added `verify_jupyter_representation`.
From the naming, it does not fit there, but I didn't want to copy that code.

xflr6 commented on graphviz/jupyter_integration.py:
```diff
@@ -1,12 +1,85 @@
+    _verify_jupyter_representation = staticmethod(verify_jupyter_representation)
+
+    @property
+    def jupyter_representation(self) -> str:
```
Do you think it is essential to set the output format also per-instance as opposed to setting a module-default that applies to all instances?

boeddeker replied:
No, I copied the tests, and they wanted this.
Since changing instance values has less side effects than module-defaults, I just added this.

xflr6 replied:
I see. I am on the fence (maybe I would call it `jupyter_format` if we want to expose it per-instance). Might adapt the tests then,

boeddeker replied:
`jupyter_repr` could also be used. While `repr` is a shortcut, it is common and each Python user should know it. Further the name is close to `_repr_svg_`.

xflr6 commented:
Thanks. Can you squash the PR into one commit?

If you are fine with me moving things, adapting naming and possibly more things, I can take over (if you have enabled it, I can add another commit to the PR branch so we can keep things separate after merging). SG?

pytype: 

```
FAILED: /home/runner/work/graphviz/graphviz/.pytype/pyi/graphviz/jupyter_integration.pyi 
/opt/hostedtoolcache/Python/3.9.8/x64/bin/python -m pytype.single --imports_info /home/runner/work/graphviz/graphviz/.pytype/imports/graphviz.jupyter_integration.imports --module-name graphviz.jupyter_integration -V 3.9 -o /home/runner/work/graphviz/graphviz/.pytype/pyi/graphviz/jupyter_integration.pyi --analyze-annotated --gen-stub-imports --nofail --quick /home/runner/work/graphviz/graphviz/graphviz/jupyter_integration.py
File "/home/runner/work/graphviz/graphviz/graphviz/jupyter_integration.py", line 60, in JupyterSvgIntegration: Type annotation for include does not match type of assignment [annotation-type-mismatch]
  Annotation: list (Did you mean 'typing.Optional[list]'?)
  Assignment: None
File "/home/runner/work/graphviz/graphviz/graphviz/jupyter_integration.py", line 60, in JupyterSvgIntegration: Type annotation for exclude does not match type of assignment [annotation-type-mismatch]
  Annotation: list (Did you mean 'typing.Optional[list]'?)
  Assignment: None
```

boeddeker commented:
> Thanks. Can you squash the PR into one commit?

Yes, I squashed now all changes.

> If you are fine with me moving things, adapting naming and possibly more things, I can take over (if you have enabled it, I can add another commit to the PR branch so we can keep things separate after merging). SG?

Yes, you can take over. I fixed the typing you mentioned and renamed `JupyterSvgIntegration` to `JupyterIntegration` in my last commit.
